src/libp2p_crypto.erl

```diff
@@ -370,10 +500,218 @@ base58check_decode(B58) ->
             {error, bad_checksum}
     end.
 
+-spec pubkey_is_multisig(pubkey()) -> boolean().
+pubkey_is_multisig({multisig, _, _, _}) ->
+    true;
+pubkey_is_multisig({ecc_compact, _}) ->
+    false;
+pubkey_is_multisig({ed25519, _}) ->
+    false.
+
+%% @doc A multisig pubkey is simply an order-preserving concatanation of
+%% individual pubkey binaries.
+%% @end
+-spec make_multisig_pubkey(pos_integer(), pos_integer(), [pubkey()], HashType) ->
+    {ok, binary()} | {error, Error} when
+    Error :: {contains_multisig_keys, [pubkey()]} | term(),
+    HashType :: atom(). % FIXME Dialyzer doesn't see multihash:hash()
+make_multisig_pubkey(M, N, PubKeys, HashType) ->
+    case lists:filter(fun pubkey_is_multisig/1, PubKeys) of
+        [_|_]=PKs ->
+            {error, {contains_multisig_keys, PKs}};
+        [] ->
+            PubKeysBins = lists:map(fun pubkey_to_bin/1, PubKeys),
+            PubKeysBin = iolist_to_binary(PubKeysBins),
+            case multihash:digest(PubKeysBin, HashType) of
+                {ok, KeysDigest} ->
+                    {ok, {multisig, M, N, KeysDigest}};
+                {error, _}=Error ->
+                    Error
+            end
+    end.
+
+%% @doc A multisig-signature is a concatanation of multisig-pubkey and a list
+%% of triples of individual-signatures prefixed with an index (of corresponsing
+%% individual-pubkey in the multisig-pubkey) and the length of the
+%% individual-signature:
+%%
+%% <<PubKeys/binary, Triples/binary>>
+%% where
+%% Pubkeys = <<PK0/binary, ..., PKN/binary>>
+%% Triples =
+%%   <<
+%%     0/integer, Len0/integer, Sig0/binary,
+%%     ...,
+%%     N/integer, LenN/integer, SigN/binary
+%%   >>
+%%
+%% PubKeys MUST be in the same order as when each triple was constructed.
```

Comment:
Do you mean as when the multisig address was constructed?

Reply:
Yes

Reply:
It would be good to clarify